csharp/src/Ice/Communicator.cs

```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
-
-            //
-            // If this fails, look for helper classes in the typeIdNamespaces namespace(s)
-            //
-            if (c == null && _typeIdNamespaces != null)
+            return _typeIdCache.GetOrAdd(typeId, id =>
             {
-                foreach (string ns in _typeIdNamespaces)
+                // First attempt corresponds to no cs:namespace metadata in the
+                // enclosing top-level module
+                string className = TypeToClass(id);
+                Type? classType = AssemblyUtil.FindType(className);
+
+                //
+                // If this fails, look for helper classes in the typeIdNamespaces namespace(s)
+                //
+                if (classType == null && _typeIdNamespaces != null)
                 {
-                    Type? helper = AssemblyUtil.FindType(ns + "." + className);
-                    if (helper != null)
+                    foreach (string ns in _typeIdNamespaces)
                     {
-                        try
-                        {
-                            c = helper.GetProperty("targetClass").PropertyType;
-                            break; // foreach
-                        }
-                        catch (System.Exception)
+                        Type? helper = AssemblyUtil.FindType(ns + "." + className);
+                        if (helper != null)
                         {
+                            try
+                            {
+                                classType = helper.GetProperty("targetClass").PropertyType;
+                                break; // foreach
+                            }
+                            catch (System.Exception)
+                            {
+                            }
                         }
                     }
                 }
-            }
 
-            //
-            // Ensure the class is instantiable.
-            //
-            if (c != null && !c.IsAbstract && !c.IsInterface)
-            {
-                return c;
-            }
+                //
+                // Ensure the class is instantiable.
+                //
+                if (classType != null && !classType.IsAbstract && !classType.IsInterface)
+                {
+                    return classType;
+                }
 
-            return null;
+                return null;
+            });
         }
 
-        internal string? ResolveCompactId(int compactId)
+        internal Type? ResolveCompactId(int compactId)
         {
-            string[] defaultVal = { "IceCompactId" };
-            var compactIdNamespaces = new List<string>(defaultVal);
-
-            if (_typeIdNamespaces != null)
+            return _compactIdCache.GetOrAdd(compactId, id =>
             {
-                compactIdNamespaces.AddRange(_typeIdNamespaces);
-            }
+                string[] defaultVal = { "IceCompactId" };
+                var compactIdNamespaces = new List<string>(defaultVal);
```

Comment:
Should we cache compacIdNamespaces like we cache typeIdNamespaces?

Reply:
I think we should!